### PR TITLE
allow passing of docker build options and document dockerfile deployment. closes #1231

### LINKED ISSUE #1231: Docker steps are not cached
When using Dockerfiles, each build step is not cached. This looks to be by design:
https://github.com/progrium/dokku/commit/9a71fff3f1e599e4148b9562b092c7388018add1

Is there a way of allowing the occasional forcing of fresh builds without disabling caching for every build step ?

My particular use-case is shown here, where the package restore is rather slow and mostly the same each time:
https://github.com/aspnet/aspnet-docker/issues/59

Maybe Dokku could include a http proxy for this kind of thing ? Or a cache volume could be attached automatically ?

### PR DESCRIPTION
**Pull request by michaelshobbs**
- documents Dockerfile deployments
- exposes `$DOKKU_DOCKERFILE_CACHE_BUILD` for specific control of `--no-cache` on `docker build`
- exposes `$DOKKU_DOCKER_BUILD_OPTS` for injection of any `docker build` option
- sources `$DOKKU_ROOT/$APP/ENV` on `dokku build`

### CODE REVIEW

File: docs/application-deployment.md [hunk 1]
```diff
@@ -50,6 +50,14 @@ ssh -T git@github.com
 
 If buildpack detection isn't working well for you or you want to specify a custom buildpack for one repository you can create & commit a file in the root of your git repository named `.env` containing `export BUILDPACK_URL=<repository>` before pushing. This will tell buildstep to fetch the specified buildpack and use it instead of relying on the built-in buildpacks & their detection methods.
 
+## Dockerfile deployment
+
+Deployment of `Dockerfile` repos is supported as of v0.3.15. Simply place a Dockerfile in the root of your repo and push to dokku. If you are converting from a heroku/dokku buildpack deployment, ensure you are not setting `$BUILDPACK_URL` or including a `.buildpacks` file in the root of your repo.
+
+By default, we will extract the first `EXPOSE` port and tell nginx to proxy your app to that port. Alternatively, you can set `$DOCKERFILE_PORT` in your app's dokku configuration.
+
+You can also pass any `docker build` options to `docker`, by setting `$DOKKU_DOCKER_BUILD_OPTS`. Lastly, control of docker's image layer cache when deploying Dockerfiles with dokku can be had specifically by setting `$DOKKU_DOCKERFILE_CACHE_BUILD` to `true` or `false`.
```
Comment (Flink) [inline docs/application-deployment.md]: `can be had` => typo?

**Comment by Flink** [comment]
Is `DOKKU_DOCKERFILE_CACHE_BUILD` really useful since `DOKKU_DOCKER_BUILD_OPTS` seems to be a better way to achieve the same thing (and more flexible)?

**Comment by hhff** [comment]
:+1:

**Comment by michaelshobbs** [comment]
@Flink `$DOKKU_DOCKERFILE_CACHE_BUILD` is meant to be an easy "on/off switch" and `$DOKKU_DOCKER_BUILD_OPTS` is for more "advanced" purposes. The former seems easiest for those not wanting to know about the specifics of the `docker` cli.

**Comment by Flink** [comment]
@michaelshobbs Ok works for me :wink: :+1:

**Pull request merged**